### ISSUE TITLE
NFS5 track parser

### ISSUE THREAD

**Issue by DmitryKiselevsky** [OpenNFS/OpenNFS#29]
Where can i see the NFS5 track format and/or code of parser?

**Comment by dima424658** [comment]
Ive just imported alps track to blender!
You need to decompress map with [crp2drp tool](http://www.cgig.ru/forum/viewtopic.php?t=2673) (you can also use [this](https://github.com/OpenNFS/OpenNFS/blob/a3c928077add2bd5d5f3463daee6008f150a0a54/src/Util/Utils.cpp#L241)) and just load it with [`NFS5::LoadCRP` function](https://github.com/OpenNFS/OpenNFS/blob/a3c928077add2bd5d5f3463daee6008f150a0a54/src/Loaders/NFS5/NFS5Loader.cpp#L58)

**Comment by AmrikSadhra** [comment]
Yep that's it! Nice job Dima! Sorry for the delay. There is just some extremely hacky code I wrote to bridge Arushan's [CRPLib](https://github.com/OpenNFS/OpenNFS/tree/main/include/CrpLib) to the OpenNFS format. Walk the Crp structure returned for geometry. Here is the [CRP Format](https://github.com/OpenNFS/OpenNFS/files/8219001/NFSPUCrp.txt) written by Addict and Jesper. 

Somewhere I have a more complete loader for the Crp data, with indices and texture loading, though I'm struggling to find it on any of my local branches - it's been a while, I'll commit it when found.

**Comment by dima424658** [comment]
Thanks for reply
Ive tried to write a LZ77 decompressor, but nothing came out. Maybe the source code of CRP2DRP is left somewhere?

**Comment by Fallenleader** [comment]
Having this whipped into a small hacky command line tool would be godly. Dropping in the DRP result and getting a FBX is sufficient for an early result, as we can already extract textures and assign them in something like Unity or Blender by hand if needed.
My end goals are aiming to use substance designer and painter to just outright retexture the entire track anyhow, though I could see why someone would want the OG textures eventually.

**Comment by dima424658** [comment]
I'm planning to separate parsers to LibOpenNFS and refactor them. For now i moved to C++17 and removed heavy dependencies like boost.
I don't have much free time, but I'll try to make a utility for you, [Fallenleader](https://github.com/Fallenleader)

Also i have quite a lot of work on reverse of Assetto Corsa, it would be cool to use AC physics engine somehow in this project or maybe make a mod for AC.

**Comment by Fallenleader** [comment]
Not super sure about what AC is, was planning to scour over this project to see how you are handling driving logic though.
Shy of a full reversed source for NFS3/4 for arcady physics, and NFS5 for more simulated ones, it might just be better to code some from scratch and build it up over time.
I personally and using Unity due to long time experience with the engine to bend things graphically to my will and target consoles that I admittedly have no business targeting :P
I am hoping things all end up coming together in such a way that I can craft a full retro style experience that contains all NFS1/2/3/4/5 tracks. plenty of conversions still exist out there that put NFS 1/2 tracks on 3/4, and have some potential sources for updated car models.
All I lack now is time and usable intermediate formats for the classic models, namely tracks.
A tool to do click conversion for the PC tracks and textures is super welcome, though I am neither in a rush nor wish to make you feel pressured to hurry up for such. For starters I have to get a handle on substance to really do anything with it all, otherwise it's just a asset flip to another engine.
For an example of one of the things I had in mind, I was planning to rewrite a really decent PSX shader to support modern PBR lighting and textures while still producing that texture wobble and poly jitter effect as a "retro mode" shader option the user can toggle on as a hail back to the olden days.

**Comment by dima424658** [comment]
Hi there
To be honest, I don’t really like the OpenNFS engine itself from the POV of architecture.
I would like to find some simpler and more understandable solution with decent physics and graphics, mainly for NFS:PU.
AC has debug symbols, so I took a closer look at it.
I put all the work on LibOpenNFS into [this repository](https://github.com/dima424658/LibOpenNFS). I have almost finished reading CRP files, and there is already conversion to FBX or OBJ just around the corner

**Comment by dima424658** [comment]
I'm sorry for delay, don't have much free time doing this. So the parser works perfect except UVs. They have really strange values. Maybe they need to be normalized or something like that...
Still can't find the issue

**Comment by Fallenleader** [comment]
It has been exactly a year since you commented this, for a nice laugh.
Seriously thanks for the work, will look into it. In college now, so I haven't had much time to look into or really care about side projects, but this definitely got my attention.
Edit: As I am no programmer, all I can say is great work, and hoping that a utility to convert to FBX or OBJ comes along (even a mel script for Maya 2023 would work).
UVs are less of a concern than the geometry, as for my immediate intent they will be regenerated anyhow to paint in substance, though I can see why this would be a concern for purists who want a direct 1:1

**Comment by dima424658** [comment]
I'm sorry it took me so long to respond.<br>@e-rk just made an addon for Blender - [e-rk/speedtools](https://github.com/e-rk/speedtools)
